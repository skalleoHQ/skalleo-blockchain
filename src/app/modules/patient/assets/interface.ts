export interface CreatePatientAccountAssetProps {
	patientIdentificationNumber: string;
	areaCode: string;
	username: string;
}