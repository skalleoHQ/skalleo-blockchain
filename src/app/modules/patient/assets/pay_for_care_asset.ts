import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

export class PayForCareAsset extends BaseAsset {
	public name = 'payForCare';
  public id = 2;

  // Define schema for asset
	public schema = {
    $id: 'patient/payForCare-asset',
		title: 'PayForCareAsset transaction asset for patient module',
		type: 'object',
		required: [],
		properties: {},
  };

  public validate({ asset }: ValidateAssetContext<{}>): void {
    // Validate your asset
  }

	// eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<{}>): Promise<void> {
		throw new Error('Asset "payForCare" apply hook is not implemented.');
	}
}
