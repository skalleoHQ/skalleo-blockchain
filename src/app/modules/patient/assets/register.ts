export interface PatientModuleProps {
    selfPatient: 'bytes';
    reverseLookup: 'bytes';
};


export interface CreatePatientAccountAssetProps {
	patientIdentificationNumber: string;
	areaCode: string;
	username: string;
};


export interface RegisteredPatientAccountsSchemaProps {
	registeredPatientAccounts: []
};



