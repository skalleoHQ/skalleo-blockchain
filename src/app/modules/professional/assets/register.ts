export interface ProfessionalModuleProps{
    selfProfessional: Buffer,
    reverseLookup: Buffer
};


export interface CreateProfessionalAccountAssetProps {
    readonly professionalIdentificationNumber: string;
    readonly areaCode: string;
    readonly username: string;
};


export interface RegisteredProfessionalAccountsSchemaProps {
    readonly registeredProfessionalAccounts: [];
};


export interface RegisteredProfessionalAccount {
    readonly id: Buffer;
    readonly ownerAddress: Buffer;
    readonly professionalIdentificationNumber: string;
    readonly areaCode: string;
    readonly username: string;
}




