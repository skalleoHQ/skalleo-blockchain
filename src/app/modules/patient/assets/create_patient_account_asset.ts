import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { CreatePatientAccountAssetProps } from './register';
import { PatientModuleProps } from './register';


const {
	VALID_PATIENT_DOMAIN,
	createPatientAccount,
	getAllPatientAccounts,
	setAllPatientAccounts,

} = require("../data/utils")




export class CreatePatientAccountAsset extends BaseAsset<CreatePatientAccountAssetProps> {
	public name = 'createPatientAccount';
  	public id = 1;

  // Define schema for asset
	public schema = {
    $id: 'patient/createPatientAccount-asset',
		title: 'CreatePatientAccountAsset transaction asset for patient module',
		type: 'object',
		required: ['patientIdentificationNumber', 'areaCode', 'username'],
		properties: {
			patientIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			username: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3,
			},
		},
  	};

	public validate({ asset }: ValidateAssetContext<CreatePatientAccountAssetProps>): void {

		// Verify if client fill patient Identification Number box
		if (!asset.patientIdentificationNumber) {
			throw new Error('You must enter your identification number');
		};

		// Implement checking identity procedure (redirect the client to national database for example and verify identity)

		/**
		 * */ 

		// Verify if client fill areaCode box
		if (!asset.areaCode) {
			throw new Error('You must enter your areaCode, you can find it on https://skalleo.io');
		};

		// Implement checking areaCode procedure (redirect the client to national database for example and verify if this zone is concerned)
		// OR verify it with apply: we can look CHAIN_STATE_POOL_ACCOUNTS if the areaCode is available

		/**
		 * */
		

		// Verify if username is valid
		const chunks = asset.username.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PATIENT_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid domain is "${VALID_PATIENT_DOMAIN.join()}"`);
		};
		

	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<CreatePatientAccountAssetProps>): Promise<void> {
	
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get<PatientModuleProps>(senderAddress);

		const patientAccounts = await getAllPatientAccounts(stateStore);
	
		//Each patientIdentificationNumber has only one account
		const patientIdentificationNumberIndex = patientAccounts.findIndex((t) => t.id.equals(asset.patientIdentificationNumber));
		if (patientIdentificationNumberIndex >= 0) {
			throw new Error('You have already a patient account !')
		}

		//Verify if domain is unique
		const patientUsernameIndex = patientAccounts.findIndex((t) => t.id.equals(asset.username));
		if (patientUsernameIndex >= 0) {
			throw new Error('This username is already reserved, please try another.')
		}

		//Admit only one patient account by the way
		if (senderAccount.selfPatient) {
			throw new Error('You have already a patient account !')
		}

		//Verify if areaCode is valid

		/*const areaCodeIndex = poolAccounts.findIndex((t) => t.id.equals(asset.areaCode));
		if (areaCodeIndex >= 0) {
			throw new Error('Your area is not yet concerned, sorry !');
		}*/

		//create patient account
		const patientAccount = createPatientAccount({
			patientIdentificationNumber: asset.patientIdentificationNumber,
        	areaCode: asset.areaCode,
        	username: asset.username,
        	ownerAddress: senderAddress,
			nonce: transaction.nonce,
		});

		//update user account with unique username and his personal patient account
		senderAccount.selfPatient = patientAccount;
		senderAccount.reverseLookup = patientAccount.username;
		await stateStore.account.set(senderAddress, senderAccount);
		
		//save patient in database
		const allAccounts = await getAllPatientAccounts(stateStore);
		allAccounts.push(patientAccount);
		await setAllPatientAccounts(stateStore, allAccounts);

	}

}



module.exports = {
	CreatePatientAccountAsset,
}



