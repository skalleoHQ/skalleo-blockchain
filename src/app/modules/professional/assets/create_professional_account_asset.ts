import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

import { CreateProfessionalAccountAssetProps, ProfessionalModuleProps } from './register';

const { 
	VALID_PROFESSIONAL_DOMAIN,
	getAllProfessionalAccounts,
	createProfessionalAccount,
	setAllPatientAccounts,

} = require ('../data/utils')


export class CreateProfessionalAccountAsset extends BaseAsset<CreateProfessionalAccountAssetProps> {
	public name = 'createProfessionalAccount';
  	public id = 1;

  // Define schema for asset
	public schema = {
    $id: 'professional/createProfessionalAccount-asset',
		title: 'CreateProfessionalAccountAsset transaction asset for professional module',
		type: 'object',
		required: ['professionalIdentificationNumber', 'areaCode', 'username'],
		properties: {
			professionalIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			username: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3,
			}
		},
  	};



	public validate({ asset }: ValidateAssetContext<CreateProfessionalAccountAssetProps>): void {
		// Verify if client fill professional Identification Number box
		if (!asset.professionalIdentificationNumber) {
			throw new Error('You must enter your identification number');
		};

		// Must implement checking identity procedure (redirect the client to his national database for example and verify identity)

		/**
		 * */ 

		// Verify if client fill areaCode box
		if (!asset.areaCode) {
			throw new Error('You must enter your areaCode, you can find it on https://skalleo.io');
		};

		// Implement checking areaCode procedure (redirect the client to his national database for example and verify if this zone is concerned)
		// OR verify it with apply: we can look CHAIN_STATE_POOL_ACCOUNTS if the areaCode is available

		/**
		 * */
		
		// Verify if professional fill username box
		if (!asset.username) {
			throw new Error('You must enter your username, please use "." only in front of "pro');
		};

		// Verify if username is valid
		const chunks = asset.username.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PROFESSIONAL_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid domain is "${VALID_PROFESSIONAL_DOMAIN.join()}"`);
		};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<CreateProfessionalAccountAssetProps>): Promise<void> {
		
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get<ProfessionalModuleProps>(senderAddress);

		const professionalAccounts = await getAllProfessionalAccounts(stateStore);

		//Admit only one professional account by the way
		if (senderAccount.selfProfessional || senderAccount.reverseLookup) {
			throw new Error('You have already a professional account !')
		}

		//Each professionalIdentificationNumber has only one account
		const professionalIdentificationNumberIndex = professionalAccounts.findIndex((t) => t.professionalIdentificationNumber === (asset.professionalIdentificationNumber));
		if (professionalIdentificationNumberIndex >= 0) {
			throw new Error('This professional identification number is already registered !')
		}

		//Verify if username is unique
		const professionalUsernameIndex = professionalAccounts.findIndex((t) => t.username === (asset.username));
		if (professionalUsernameIndex >= 0) {
			throw new Error('This username is already reserved, please try another.')
		}

		//Verify if areaCode is valid

		/*const areaCodeIndex = poolAccounts.findIndex((t) => t.areaCode === (asset.areaCode));
		if (areaCodeIndex >= 0) {
			throw new Error('Your area is not yet concerned, sorry !');
		}*/

		//create professional account
		const professionalAccount = createProfessionalAccount({
			professionalIdentificationNumber: asset.professionalIdentificationNumber,
        	areaCode: asset.areaCode,
        	username: asset.username,
        	ownerAddress: senderAddress,
			nonce: transaction.nonce,
		});

		//update user account with unique username and his personal professional account
		senderAccount.selfProfessional = professionalAccount;
		senderAccount.reverseLookup = professionalAccount.username;
		await stateStore.account.set(senderAddress, senderAccount);
		
		//save professional in database
		const allAccounts = await getAllProfessionalAccounts(stateStore);
		allAccounts.push(professionalAccount);
		await setAllPatientAccounts(stateStore, allAccounts);

	}

}




module.exports = {
	CreateProfessionalAccountAsset,
}
