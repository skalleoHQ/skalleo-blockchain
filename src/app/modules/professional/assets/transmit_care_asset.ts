import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { TransmitCareAssetProps } from './register';


const {
	VALID_PATIENT_DOMAIN,
	getAllPatientAccounts,

} = require ('../../patient/data/utils');



const {
	getAllProfessionalAccounts,

} = require ('./register');



const {
	recordCare,
	getAllRecordedCare,
	setAllRecordedCare,
	
} = require ('../data/care')



export class TransmitCareAsset extends BaseAsset<TransmitCareAssetProps> {
	public name = 'transmitCare';
  	public id = 2;

  // Define schema for asset
	public schema = {
    $id: 'professional/transmitCare-asset',
		title: 'TransmitCareAsset transaction asset for professional module',
		type: 'object',
		required: ['patientIdentificationNumber', 'reverseLookup', 'areaCode', 'careSpecifications'],
		properties: {
			patientIdentificationNumber: {
				dataType: 'string',
				fieldNumber: 1,
			},
			reverseLookup: {
				dataType: 'string',
				fieldNumber: 2,
			},
			areaCode: {
				dataType: 'string',
				fieldNumber: 3
			},
			careSpecifications: {
				dataType: 'string',
				fieldNumber: 4,
			},
		},
	};

	public validate({ asset }: ValidateAssetContext<TransmitCareAssetProps>): void {

		if (!asset.patientIdentificationNumber) {
			throw new Error('You must enter identification number of your patient');
		}

		if (!asset.reverseLookup) {
			throw new Error('You must enter username of your patient');
		}

		const chunks = asset.reverseLookup.split(/\./);

		if (chunks.length > 2) {
			throw new Error('You may not use "." than for the domain');
		};

		if (!VALID_PATIENT_DOMAIN.includes(chunks[1])) {
			throw new Error(`Invalid domain found "${chunks[1]}". Valid domain is "${VALID_PATIENT_DOMAIN.join()}"`);
		};

		if (!asset.areaCode) {
			throw new Error('You must enter areaCode of your patient');
		}

		//This careSpecifications will be used to update patient's medical record in his national database and plugin database
		if (!asset.careSpecifications) {
			throw new Error('Please update the cares specifications of your patient !')
		}

		//Must implement procedure in order to send specifications to national database & verify if careSpecs respect skeleton

	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<TransmitCareAssetProps>): Promise<void> {
		const senderAddress = transaction.senderAddress;
		
		//Verify if patient exists
		const patientAccounts = await getAllPatientAccounts(stateStore);

		const patientIdentificationNumberIndex = patientAccounts.findIndex((t) => t.patientIdentificationNumber === asset.patientIdentificationNumber);
		if (patientIdentificationNumberIndex < 0) {
			throw new Error('Patient not found !');	
		}

		const patientReverseLookupIndex = patientAccounts.findIndex((t) => t.username === asset.reverseLookup);
		if (patientReverseLookupIndex < 0) {
			throw new Error('Patient not found !');
		}

		/*const patientAreaCodeIndex = patientAccounts.findIndex((t) => t.areaCode === asset.areaCode);
		if (patientAreaCodeIndex < 0) {
			throw new Error('Patient not found !');
		}*/
		//assert(patientReverseLookupIndex === patientAreaCodeIndex); impossible because areaCode not unique

		if(patientIdentificationNumberIndex !== patientReverseLookupIndex) {
			throw new Error('Patient not found !');
		}

		const professionalAccounts = await getAllProfessionalAccounts(stateStore);
		const addressIndex = professionalAccounts.findIndex((t) => t.ownerAddress === transaction.senderAddress);
		
		const professionalAccount = professionalAccounts[addressIndex];
		const patientAccount = patientAccounts[patientIdentificationNumberIndex]; 

		if (professionalAccount.ownerAddress === patientAccount.ownerAddress) {
			throw new Error('Transaction error !');
		}
		
		//create care for a patient
		const care = recordCare({
			patientIdentificationNumberIndex: asset.patientIdentificationNumber,
			reverseLookup: asset.reverseLookup,
			areaCode: asset.areaCode,
			careSpecifications: asset.careSpecifications,
			senderAddress: senderAddress,
			nonce: transaction.nonce,
		})

		//update care data chain
		const allCare = await getAllRecordedCare(stateStore);
		allCare.push(care);
		await setAllRecordedCare(stateStore, allCare);

	}


}



module.exports = {
	TransmitCareAsset,
}
