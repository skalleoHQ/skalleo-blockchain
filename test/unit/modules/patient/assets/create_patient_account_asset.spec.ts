import { testing } from 'lisk-sdk';
import { CreatePatientAccountAsset } from '../../../../../src/app/modules/patient/assets/create_patient_account_asset';

describe('CreatePatientAccountAsset', () => {
  let transactionAsset: CreatePatientAccountAsset;

	beforeEach(() => {
		transactionAsset = new CreatePatientAccountAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(1);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('createPatientAccount');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});



	describe('validate', () => {
		describe('schema validation', () => {
			it('should throw error if username has more than one domain', () => {
				const context = testing.createValidateAssetContext({
					asset: {patientIdentificationNumber: "3301012022", areaCode: "33_France", username: "by.moussa.adh"},
					transaction: {senderAddress: Buffer.alloc(0)} as any,
				});
				expect(() => transactionAsset.validate(context)).toThrow(
					'You may not use "." than for the domain'
				)
			});

			it('should throw error if username has more than one domain', () => {
				const context = testing.createValidateAssetContext({
					asset: {patientIdentificationNumber: "3301012022", areaCode: "33_France", username: "by.moussa"},
					transaction: {senderAddress: Buffer.alloc(0)} as any,
				});
				expect(() => transactionAsset.validate(context)).toThrow(
					'Invalid domain found "moussa". Valid domain is "adh"'
				)
			});




			it.todo('should be ok for valid schema');
		});
	});



	describe('apply', () => {
		describe('valid cases', () => {
			it.todo('should update the state store');
		});


		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
