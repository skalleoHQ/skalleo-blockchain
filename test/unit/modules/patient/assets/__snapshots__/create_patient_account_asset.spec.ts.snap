// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePatientAccountAsset constructor should have valid schema 1`] = `
Object {
  "$id": "patient/createPatientAccount-asset",
  "properties": Object {
    "areaCode": Object {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "patientIdentificationNumber": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
    "username": Object {
      "dataType": "string",
      "fieldNumber": 2,
    },
  },
  "required": Array [
    "patientIdentificationNumber",
    "areaCode",
    "username",
  ],
  "title": "CreatePatientAccountAsset transaction asset for patient module",
  "type": "object",
}
`;
